File: core/src/main/java/com/tek/core/TekCoreProperties.java
package com.tek.core;

import com.tek.core.properties.TekCorsProperties;
import com.tek.core.properties.TekMailProperties;
import com.tek.core.properties.TekSchedulerProperties;
import com.tek.core.properties.file.TekFileProperties;
import com.tek.core.properties.i18n.TekLocaleProperties;
import lombok.Data;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.context.annotation.Configuration;

import static com.tek.core.TekCoreConstant.TEK_CORE;


/**
 * Tek Core Module properties to be evaluated from application.yaml / application.properties files.
 *
 * @author MarcoPagan
 */
@Configuration
@ConfigurationProperties(prefix = TEK_CORE)
@Data
public class TekCoreProperties {

    private TekCorsProperties cors = new TekCorsProperties();
    private TekLocaleProperties locale = new TekLocaleProperties();
    private TekFileProperties file = new TekFileProperties();
    private TekMailProperties mail = new TekMailProperties();
    private TekSchedulerProperties scheduler = new TekSchedulerProperties();
}

File: core/src/main/kotlin/com/tek/core/TekCoreProperties.kt
package com.tek.core

import com.tek.core.configuration.TekCorsFilter
import com.tek.core.util.LabelEnum
import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.context.annotation.Configuration
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.servlet.config.annotation.InterceptorRegistry

@Configuration
@ConfigurationProperties(prefix = TEK_MODULE_CORE)
class TekCoreProperties {
    val cors: TekCorsProperties = TekCorsProperties()
    val runner: TekRunnerProperties = TekRunnerProperties()
    val locale: TekLocaleProperties = TekLocaleProperties()
}

/**
 * Configuration properties for [TekCorsFilter].
 *
 * Fallback to default configuration if none provided.
 */
class TekCorsProperties {
    internal enum class RequestHeader(
        override val label: String
    ) : LabelEnum {
        X_REQUESTED_WITH("x-requested-with"),
        L_AUTHORIZATION("authorization"),
        CONTENT_TYPE("Content-Type"),
        U_AUTHORIZATION("Authorization"),
        CREDENTIAL("credential"),
        X_XSRF_TOKEN("X-XSRF-TOKEN")
    }

    var allowedOrigin: String = "http://localhost:4200"
    var allowedCredentials: String = "true"
    var allowedMethods: Array<String> =
        RequestMethod.values().map { it.name }.toTypedArray()
    var allowedHeaders: Array<String> =
        RequestHeader.values().map { it.label }.toTypedArray()
}

/**
 * Configuration properties for [TekDataRunner].
 */
class TekRunnerProperties {
    enum class TekRunnerAction { CREATE, NONE }

    var action: TekRunnerAction = TekRunnerAction.NONE
}

/**
 * Configuration properties for [InterceptorRegistry.addInterceptor] to manage app locale resolution.
 *
 * Fallback to default configuration if none provided.
 */
class TekLocaleProperties {
    /** Session, Cookie */
    enum class TekLocaleType { SESSION, COOKIE, ACCEPT_HEADER }

    var type: TekLocaleType = TekLocaleType.SESSION
    var cookieName: String = "locale"
    var cookieMaxAge: Int = 24 * 60 * 60
    var interceptorName: String = "lang"
}





